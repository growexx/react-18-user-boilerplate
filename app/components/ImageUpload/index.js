/**
 *
 * ImageUpload
 *
 */

import React, { useState } from 'react';
import { Upload } from 'antd';
import ImgCrop from 'antd-img-crop';
import 'antd/es/modal/style';
import 'antd/es/slider/style';
import { API_ENDPOINTS } from 'containers/constants';

function ImageUpload() {
  const [fileList, setFileList] = useState([]);

  const onChange = ({ fileList: newFileList }) => {
    setFileList(newFileList);
  };

  const onPreview = async file => {
    let src = file.url;
    if (!src) {
      src = await new Promise(resolve => {
        const reader = new FileReader();
        reader.readAsDataURL(file.originFileObj);
        reader.onload = () => resolve(reader.result);
      });
    }
    const image = new Image();
    image.src = src;
    const imgWindow = window.open(src);
    imgWindow.document.write(image.outerHTML);
  };

  return (
    <ImgCrop rotationSlider>
      <Upload
        action={API_ENDPOINTS.IMAGE_UPLOAD}
        listType="picture-card"
        fileList={fileList}
        onChange={onChange}
        onPreview={onPreview}
        maxCount={1}
      >
        {fileList.length < 1 && '+ Upload'}
      </Upload>
    </ImgCrop>
  );
}

export default ImageUpload;
