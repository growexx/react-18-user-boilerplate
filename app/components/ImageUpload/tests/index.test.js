/**
 *
 * Tests for ImageUpload
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import userEvent from '@testing-library/user-event';
import ImageUpload from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';
import getCroppedImg from '../util/crop';
import image from './test.jpg';

jest.mock('../util/crop');

const imageObjectUrl = 'https://picsum.photos/1000';
const uploadFile = async (file, container) => {
  const fileInput = container.querySelector('input');

  await userEvent.upload(fileInput, file);
  return fileInput;
};

global.URL.createObjectURL = jest.fn().mockImplementation(() => imageObjectUrl);

describe('<ImageUpload />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ImageUpload />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should handle image crop modal close', async () => {
    const { container, getByText } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ImageUpload />
      </IntlProvider>,
    );
    const file = new File([image], 'test.png', {
      type: 'image/png',
      lastModified: Date.now(),
    });

    await uploadFile(file, container);
    expect(document.querySelectorAll('.ant-modal-mask')).toHaveLength(1);

    await act(async () => {
      fireEvent.click(getByText('Cancel'));
    });
    expect(document.querySelectorAll('.ant-modal-mask')).toHaveLength(0);
  });

  it('should change image zoom and rotation', async () => {
    const { container } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ImageUpload />
      </IntlProvider>,
    );
    const file = new File([image], 'test.png', {
      type: 'image/png',
      lastModified: Date.now(),
    });

    await uploadFile(file, container);

    const zoomHandle = document.querySelectorAll('.ant-slider-handle')[0];
    const rotationHandle = document.querySelectorAll('.ant-slider-handle')[1];

    fireEvent.focus(zoomHandle);
    fireEvent.keyDown(zoomHandle, {
      key: 'ArrowRight',
      code: 'ArrowRight',
      keyCode: 39,
      charCode: 39,
    });

    fireEvent.focus(rotationHandle);
    fireEvent.keyDown(rotationHandle, {
      key: 'ArrowRight',
      code: 'ArrowRight',
      keyCode: 39,
      charCode: 39,
    });

    expect(zoomHandle.getAttribute('aria-valuenow')).not.toBe(
      zoomHandle.getAttribute('aria-valuemin'),
    );
    expect(rotationHandle.getAttribute('aria-valuenow')).not.toBe(
      rotationHandle.getAttribute('aria-valuemin'),
    );
  });

  it('should handle image crop', async () => {
    getCroppedImg.mockImplementation(() => Promise.resolve(imageObjectUrl));
    const { container, getByText } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <ImageUpload />
      </IntlProvider>,
    );
    const file = new File([image], 'test.png', {
      type: 'image/png',
      lastModified: Date.now(),
    });

    await uploadFile(file, container);
    expect(document.querySelectorAll('.ant-modal-mask')).toHaveLength(1);

    await act(async () => {
      fireEvent.click(getByText('OK'));
    });
    expect(document.querySelectorAll('.ant-modal-mask')).toHaveLength(0);
  });
});
