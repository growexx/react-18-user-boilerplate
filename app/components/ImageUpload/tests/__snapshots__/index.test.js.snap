// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageUpload /> Should render and match the snapshot 1`] = `
<span
  class="ant-upload-wrapper css-dev-only-do-not-override-byeoj0 ant-upload-picture-card-wrapper"
>
  <div
    class="ant-upload-list ant-upload-list-picture-card"
  >
    <div
      class="ant-upload ant-upload-select"
    >
      <span
        class="ant-upload"
        role="button"
        tabindex="0"
      >
        <input
          accept="image/*"
          style="display: none;"
          type="file"
        />
        + Upload
      </span>
    </div>
  </div>
</span>
`;
